vllm/inputs/data.py

```diff
@@ -163,6 +163,14 @@ class TokenInputs(TypedDict):
     to pass the mm_processor_kwargs to each of them.
     """
 
+    instruction_seq: NotRequired[Optional[str]]
+    """
+    The instruction sequence that is usually prepended to the original prompt
+    when passing to the model. Certain models need to extract this instruction
+    sequence from the prompt in order to adjust certain operations of the
+    model such as the attention mask.
+    """
```

Comment:
Hmm, I'm not a fan of adding this just for one model. It would be better to pass this as multi-modal inputs, but that's currently being refactored.

Reply:
it should be possible to get rid of this by introducing special tokens (similar to <|im_begin|>, <|im_end|>). we'll give it a try and update the PR.

Reply:
@DarkLight1337 The only reason I have added `instruction_seq` to data.py and scheduler.py is that I need to extract the `instruction_seq` from the prompt and then use it in the pooler. If we want to do the extraction in `InputProcessor`, then I would need to somehow include the `instruction_seq` in the output of `InputProcessor` and then pass that to the pooler somehow. This is what I have done in the PR. I can switch to using `multi_modal_inputs` but I would still need to pass that somehow to the pooler which would require a similar scheduler change.

An alternative approach is to do the extraction inside the pooler. The good news is that we have access to the `prompt_token_ids` as part of `PoolingMetadata` which I can use for extraction.

Please let me know what you think.

Reply:
I removed the input processor and now doing pattern matching inside the pooler. So no change in data.py or scheduler.py.